### PR TITLE
adding new __replaceFunctionImplementation_unsafe built-in

### PR DESCRIPTION
**Pull request by NTillmann**
Release notes: providing __replaceFunctionImplementation_unsafe built-in

This new built-in allows replacing the method implementation (capture environment, body, ...) of source functions.

As a result, all method calls executed by the Prepack interpreter will be redirected,
and the replacement will carry over to the prepacked code.

### CODE REVIEW

File: src/intrinsics/prepack/global.js [hunk 1]
```diff
@@ -551,6 +551,58 @@ export default function(realm: Realm): void {
     })
   );
 
+  // Helper function that replaces the body of a source function with another source function body.
+  // Note that this function affects un-tracked state, so care must be taken
+  // that this helper function is executed at the right time.
```
Comment (hermanventer) [inline src/intrinsics/prepack/global.js]: What is the right time? What is the wrong time?

File: src/intrinsics/prepack/global.js [hunk 2]
```diff
@@ -551,6 +551,58 @@ export default function(realm: Realm): void {
...
+            );
+          }
+          if (!(source instanceof ECMAScriptSourceFunctionValue)) {
+            throw realm.createErrorThrowCompletion(
+              realm.intrinsics.TypeError,
+              "second argument is not a function with source code"
+            );
+          }
+
+          // relevant properties for functionValue
```
Comment (hermanventer) [inline src/intrinsics/prepack/global.js]: Is it safe in general to just update all of these properties?

File: src/intrinsics/prepack/global.js [hunk 3]
```diff
@@ -551,6 +551,58 @@ export default function(realm: Realm): void {
     })
   );
 
+  // Helper function that replaces the body of a source function with another source function body.
```
Comment (hermanventer) [inline src/intrinsics/prepack/global.js]: You are not just replacing the body but the parameters too and much else beside.

**Comment by hermanventer** [comment]
This seems simple enough, but I am lost trying to figure out why you need it. I'm also not quite sure it is correct. I don't have any particular reason to believe it is incorrect, but it is not OBVIOUSLY correct either.

File: src/intrinsics/prepack/global.js (hunk 3 above)
Comment (NTillmann) [reply src/intrinsics/prepack/global.js]: The body and everything that the body may depend on. Do you have a better name in mind?

File: src/intrinsics/prepack/global.js (hunk 2 above)
Comment (NTillmann) [reply src/intrinsics/prepack/global.js]: I am not entirely sure, but it's fine for my purposes...

File: src/intrinsics/prepack/global.js (hunk 1 above)
Comment (NTillmann) [reply src/intrinsics/prepack/global.js]: That depends on what one wants. In my scenario, I'll do this replacement before the function in question will be invoked for the first time. (Ideally, I'd just replace the function object itself and be done with it without the need for `__replaceFunctionBody`; but unfortunately, the function is  stored in a way that I can't update the reference to it.)

**Pull request closed without merging**